src/Chainweb/BlockHeaderDB/PruneForks.hs

```diff
@@ -41,180 +55,199 @@ import System.LogLevel
-    !marked <- stToIO $ BF.new (BF.cheapHashes hashNum) size
+pruneForks_ logg _ (MaxRank (Max mar)) _  _
+    | mar <= 1 = 0 <$ logg Warn ("Skipping database prunning for max bound of " <> sshow mar)
+pruneForks_ logg cdb mar mir callback = do
+    logg Debug $ "Prunning block header database for chain " <> sshow (_chainId cdb)
+        <> " with upper bound " <> sshow (_getMaxRank mar)
+        <> " and lower bound " <> sshow (_getMinRank mir)
+
+    !pivots <- entries cdb Nothing Nothing (Just $ MinRank $ Min $ _getMaxRank mar) (Just mar)
+        $ fmap (force . L.nub) . S.toList_ . S.map _blockParent
+            -- the set of pivots is expected to be very small. In fact it is
+            -- almost always a singleton set.
+
+    if null pivots
+      then do
+        logg Warn
+            $ "Skipping database pruning because of an empty set of block headers at upper pruning bound " <> sshow mar
+        return 0
+      else
+        withReverseHeaderStream cdb (mar - 1) mir
+            $ S.foldM_ go (return (pivots, int (_getMaxRank mar), 0)) (\(_,_,!n) -> evaluate n)
 
-    -- Bloom filter for marking block payload hashes
-    --
-    -- Payloads can be shared between different blocks. Thus, to be on the safe
-    -- side, we are not deleting any marked payload.
-    --
-    -- Note that we could use cuckoo hashes or counting bloom filters to do
-    -- reference counting and collect more unreferenced payloads.
-    --
-    -- TODO: would it be better to use a single shared filter?
-    --
-    let (psize, phashNum) = BF.suggestSizing s 0.0001
-    !markedPayloads <- stToIO $ BF.new (BF.cheapHashes phashNum) psize
+  where
+    go :: ([BlockHash], BlockHeight, Int) -> BlockHeader -> IO ([BlockHash], BlockHeight, Int)
+    go ([], _, _) cur = throwM $ InternalInvariantViolation
+        $ "PrunForks.pruneForks_: no pivots left at block " <> encodeToText (ObjectEncoded cur)
+    go (!pivots, !prevHeight, !n) !cur
+        -- This checks the consistency of the database.
+        | prevHeight /= curHeight && prevHeight /= curHeight + 1 =
+            throwM $ InternalInvariantViolation
+                $ "PrunForks.pruneForks_: detected a corrupted database. Some block headers are missing"
+                <> ". Current pivots: " <> encodeToText pivots
+                <> ". Current header: " <> encodeToText (ObjectEncoded cur)
+                <> ". Previous height: " <> sshow prevHeight
+        | _blockHash cur `elem` pivots = do
+            callback False cur
+            let !pivots' = force $ L.nub $ _blockParent cur : L.delete (_blockHash cur) pivots
+            return (pivots', curHeight, n)
+        | otherwise = do
+            deleteHdr cur
+            callback True cur
+            return (pivots, curHeight, n+1)
+      where
+        curHeight = _blockHeight cur
+
+    deleteHdr k = do
+        -- TODO: make this atomic (create boilerplate to combine queries for
+        -- different tables)
```

Comment:
How would we do this, does our CAS solution allow for atomic multi-table deletes?

Reply:
No, the current CAS instance doesn't support this. RocksDb, however, does support it.

This `TODO` isn't related to this PR. The code existed since the very first version. It's not critical.